py/server/tests/test_pandas.py

```diff
@@ -165,6 +165,41 @@ def test_to_table_category(self):
         df2 = to_pandas(table)
         self.assertTrue(np.array_equal(df2["A"].values, df2["B"].values))
 
+    def test_to_table_nullable(self):
+        boolean_array = pd.array([True, False, None], dtype=pd.BooleanDtype())
+        int8_array = pd.array([1, 2, None], dtype=pd.Int8Dtype())
+        int16_array = pd.array([1, 2, None], dtype=pd.Int16Dtype())
+        int32_array = pd.array([1, 2, None], dtype=pd.Int32Dtype())
+        int64_array = pd.array([1, 2, None], dtype=pd.Int64Dtype())
+        float_array = pd.array([1.1, 2.2, None], dtype=pd.Float32Dtype())
+        double_array = pd.array([1.1, 2.2, None], dtype=pd.Float64Dtype())
+        string_array = pd.array(["s11", "s22", None], dtype=pd.StringDtype())
+
+        df = pd.DataFrame({
+                              "NullableBoolean": boolean_array,
+                              "NullableInt8": int8_array,
+                              "NullableInt16": int16_array,
+                              "NullableInt32": int32_array,
+                              "NullableInt64": int64_array,
+                              "NullableFloat": float_array,
+                              "NullableDouble": double_array,
+                              "NullableString": string_array,
+                          })
+
+        table = to_table(df)
+
+        self.assertIs(table.columns[0].data_type, dtypes.bool_)
+        self.assertIs(table.columns[1].data_type, dtypes.int8)
+        self.assertIs(table.columns[2].data_type, dtypes.int16)
+        self.assertIs(table.columns[3].data_type, dtypes.int32)
+        self.assertIs(table.columns[4].data_type, dtypes.int64)
+        self.assertIs(table.columns[5].data_type, dtypes.float32)
+        self.assertIs(table.columns[6].data_type, dtypes.double)
+        self.assertIs(table.columns[7].data_type, dtypes.string)
+        self.assertEqual(table.size, 3)
+        table_string = table.to_string()
+        self.assertEqual(8, table_string.count("null"))
+
```

Comment:
What happens if a user has a table like we saw, where a column is `object` and contains strings plus NaNs for missing values?  Are we going to force users to convert to `string` columns?  Yes is probably ok.  I'm just trying to understand the thinking.

Reply:
Yes, Attempting to eagerly infer data type probably has more drawbacks than benefit. Also it is reasonable to ask the users to provide specific data types if they know what they are and don't want to see PyObject columns.